# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'/Users/ewilson/DataScience/Springboard/DS Course/Data Wrangling'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv('data/step3_output.csv', index_col = 0)
df.head()

Name       state  summit_elev  vertical_drop  \
2               Hilltop Ski Area      Alaska         2090            294   
4            Sunrise Park Resort     Arizona        11100           1800   
5  Yosemite Ski & Snowboard Area  California         7800            600   
8         Boreal Mountain Resort  California         7700            500   
9                    Dodge Ridge  California         8200           1600   

   fastQuads  quad  triple  double  surface  total_chairs  ...  \
2          0     0       1       0        2             3  ...   
4          1     2       3       1        0             7  ...   
5          0     0       1       3        1             5  ...   
8          1     1       3       1        2             8  ...   
9          0     1       2       5        4            12  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
2               30.0            30.0        150.000000       36.0   
4              800.0            80.0        115.000000       49.0   
5               88.0             0.0        110.000000       84.0   
8              380.0           200.0        150.000000       54.0   
9              862.0             0.0        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
2             69.0          30.0      34.00000              152.0   
4            250.0          74.0      78.00000              104.0   
5            300.0          47.0      47.00000              107.0   
8            400.0          49.0      64.16681              150.0   
9            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
2            30.0         0  
4            80.0         1  
5             0.0         1  
8           200.0         1  
9             0.0         1  

[5 rows x 23 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
df = pd.concat([df.drop('state', axis = 1), pd.get_dummies(df.state)], axis = 1)

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardizes the scale of the features of the dataframe except the name of the resort which we don't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis = 1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = np.ravel(y)

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [7]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [8]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = lm.predict(X_test)

In [9]:
print(y.mean())

56.927323639622024


In [10]:
lm.coef_

array([-1.26050422e-01,  3.11442203e+00, -4.12031111e+12, -4.92977299e+12,
       -8.65776698e+12, -1.05994218e+13, -1.01434447e+13,  1.93072806e+13,
        2.20329317e-01,  1.77880859e+00,  9.00146484e-01, -1.60058594e+00,
        2.08984375e-01,  1.10717773e+00, -3.92578125e-01,  2.02832031e+00,
        9.14624023e+00, -2.15429688e+00,  6.87011719e-01,  1.36328125e+00,
       -2.48077890e+12, -2.48077890e+12, -8.31920032e+12, -5.48342733e+12,
       -5.48342733e+12, -5.98920420e+12, -4.27221120e+12, -3.49831294e+12,
       -4.27221120e+12, -5.48342733e+12, -2.48077890e+12, -4.91884632e+12,
       -9.21568459e+12, -6.87493615e+12, -3.49831294e+12, -4.91884632e+12,
       -2.48077890e+12, -7.98927462e+12,  1.87892084e+06, -5.98920420e+12,
       -1.04748143e+13, -4.27221120e+12, -4.91884632e+12, -4.27221120e+12,
       -7.98927462e+12, -2.48077890e+12, -2.48077890e+12, -2.48077890e+12,
       -4.27221120e+12, -6.87493615e+12, -3.49831294e+12, -4.91884632e+12,
       -4.27221120e+12, -

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [12]:
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)

In [13]:
print(evs)

-1.3006984602263551e+23


In [14]:
print(mae)

750117599547.7969


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [15]:
print(lm.intercept_)

187529399941.681


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [16]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
# pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

                   LR Coefficient
total_chairs         1.930728e+13
double               1.059942e+13
New York             1.047481e+13
surface              1.014344e+13
Michigan             9.215685e+12
triple               8.657767e+12
California           8.319200e+12
Wisconsin            7.989275e+12
Pennsylvania         7.989275e+12
New Hampshire        7.989275e+12
Minnesota            6.874936e+12
Vermont              6.874936e+12
Idaho                5.989204e+12
New Mexico           5.989204e+12
Wyoming              5.483427e+12
Maine                5.483427e+12
Colorado             5.483427e+12
Connecticut          5.483427e+12
quad                 4.929773e+12
Montana              4.918846e+12
Washington           4.918846e+12
Ohio                 4.918846e+12
Massachusetts        4.918846e+12
Oregon               4.272211e+12
Utah                 4.272211e+12
Iowa                 4.272211e+12
Illinois             4.272211e+12
West Virginia        4.272211e+12
North Carolina

<font color='red'> In this analysis, states were among the top ten but did not completely comprise the top ten. However, further sensitivity analysis revealed that dropping the resorts from Tennessee and New Jersey eliminated the very high EVS and MAE values as well as the unrealistic intercept value. 

In [17]:
np.where(X.Tennessee == 1)

(array([139]),)

In [18]:
X = X.drop(X.index[139])
y = np.delete(y, 139)
X = X.drop('Tennessee', axis = 1)

In [25]:
np.where(X['New Jersey'] == 1)

(array([89]),)

In [26]:
X = X.drop(X.index[89])
y = np.delete(y, 89)
X = X.drop('New Jersey', axis = 1)

In [27]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(evs, mae)
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

0.7268700581291341 5.712543768294177
                   LR Coefficient
New York             2.446011e+13
Michigan             2.152421e+13
California           1.943261e+13
Wisconsin            1.866264e+13
Pennsylvania         1.866264e+13
New Hampshire        1.866264e+13
Vermont              1.606135e+13
Minnesota            1.606135e+13
Idaho                1.399308e+13
New Mexico           1.399308e+13
Wyoming              1.281184e+13
Maine                1.281184e+13
Colorado             1.281184e+13
Connecticut          1.281184e+13
Montana              1.149311e+13
Ohio                 1.149311e+13
Washington           1.149311e+13
Massachusetts        1.149311e+13
Oregon               9.982555e+12
Utah                 9.982555e+12
Iowa                 9.982555e+12
Illinois             9.982555e+12
West Virginia        9.982555e+12
North Carolina       9.982555e+12
Indiana              8.174520e+12
Missouri             8.174520e+12
Virginia             8.174520e+12
total_chair

In [28]:
print(lm.intercept_)

56.22784544168519


<font color='red'> Here, states comprise the top ten in coefficients. Furthermore, dropping 'summit elev' and 'fastQuads' produced coefficients on the same order. Now, states are no longer in the top ten. 

In [31]:
#X = X.drop('summit_elev', axis = 1)
X = X.drop('fastQuads' , axis = 1)

In [32]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(evs, mae)
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

0.7264982780770959 5.710387777104386
                   LR Coefficient
AdultWeekday             9.206597
vertical_drop            2.083710
projectedDaysOpen        1.945219
Idaho                    1.932505
total_chairs             1.833699
averageSnowfall          1.778894
Virginia                 1.706084
North Carolina           1.680652
West Virginia            1.634301
TerrainParks             1.630626
Montana                  1.518044
New Hampshire            1.502989
Massachusetts            1.471986
Wyoming                  1.463871
surface                  1.448968
clusters                 1.394460
Pennsylvania             1.239540
quad                     1.224517
SkiableTerrain_ac        1.173744
Oregon                   1.044655
daysOpenLastYear         1.023171
Rhode Island             0.929927
Utah                     0.885527
Runs                     0.845167
New Mexico               0.837335
South Dakota             0.768940
California               0.670837
LongestRun_

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [51]:
X = df.drop(['Name','AdultWeekend'], axis = 1)
X = X.drop(X.loc[:,'Alaska':'Wyoming'], axis = 1)
y = df.AdultWeekend
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(evs, mae)
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

0.5199029021516768 7.534344353225305
                   LR Coefficient
AdultWeekday             9.798295
clusters                 3.563150
summit_elev              3.039686
vertical_drop            2.770467
TerrainParks             2.458459
quad                     1.971225
projectedDaysOpen        1.779138
daysOpenLastYear         1.708601
SkiableTerrain_ac        1.231396
Snow Making_ac           1.161605
surface                  1.156217
LongestRun_mi            0.843377
averageSnowfall          0.593908
yearsOpen                0.541296
fastQuads                0.396649
double                   0.359271
total_chairs             0.321779
triple                   0.321172
Runs                     0.270724
NightSkiing_ac           0.204499


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

<font color='red'> There is no variable `base_elev`. However, we drop `summit_elev` owing to the high correlation with `vertical_drop`. 

#### Model 3

In [47]:
X = X.loc[:,'vertical_drop':'clusters']scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(evs, mae)
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

0.5234888253827599 7.516552245262173
                   LR Coefficient
AdultWeekday             9.663492
clusters                 2.920544
vertical_drop            2.747260
TerrainParks             2.426006
SkiableTerrain_ac        2.388305
quad                     1.890437
projectedDaysOpen        1.531200
Snow Making_ac           1.393790
averageSnowfall          1.360252
daysOpenLastYear         1.292275
LongestRun_mi            0.960063
surface                  0.807948
yearsOpen                0.702537
Runs                     0.595943
fastQuads                0.525431
total_chairs             0.423877
double                   0.405318
NightSkiing_ac           0.153651
triple                   0.069161


#### Model 4

In [57]:
X = df.drop(['Name','AdultWeekend','summit_elev'], axis = 1)
X = X.drop(X.loc[:,'Alaska':'Wyoming'], axis = 1)
y = df.AdultWeekend
X = X.drop(X.index[139])
y = y.drop(y.index[139])
X = X.drop(X.index[89])
y = y.drop(y.index[89])
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X) 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size = 0.25, random_state = 1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)
y_pred = lm.predict(X_test)
evs = explained_variance_score(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
print(evs, mae)
coef = pd.DataFrame(abs(lm.coef_), X.columns, columns=['LR Coefficient'])
print(coef.sort_values('LR Coefficient', ascending = False))

0.6704867855898777 6.793017125811121
                   LR Coefficient
AdultWeekday             9.867637
clusters                 2.381993
TerrainParks             2.257187
vertical_drop            2.017393
SkiableTerrain_ac        1.754270
quad                     1.651570
averageSnowfall          1.504968
daysOpenLastYear         1.449228
projectedDaysOpen        1.303399
LongestRun_mi            1.169990
Snow Making_ac           1.074732
surface                  0.983276
double                   0.604628
yearsOpen                0.415096
total_chairs             0.386606
triple                   0.284055
fastQuads                0.126444
Runs                     0.107195
NightSkiing_ac           0.069649


## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -1.300e+23| 7.501e+11|  |---|
| Model 1a.| 0.726| 5.710| 'summit_elev','base_elev','fastQuads',Tennessee & New Jersey obs|
| Model 2. | 0.520| 7.534| 'state'|
| Model 3. | 0.523| 7.517| 'state','summit_elev','base_elev'|| 
| Model 4. | 0.670| 6.793| 'state','summit_elev','base_elev',Tennessee & New Jersey obs|

Model Selection:

Although perhaps not the response desired here, removing the single Tennessee and New Jersey observations along with dropping the `summit_elev`, `base_elev`, and `state` features seem to provide the best results. Furthermore, as shown earlier, including the state features, with the appropriate `Tennessee` and `New Jersey` columns seem to provide the best results. 